File: 투포인터/leetcode352(Hard)-Data Stream as Disjoint Intervals.js
// 824ms(12.50%), 61MB(12.50%)
var SummaryRanges = function () {
  this.arr = new Set()
}

/**
 * @param {number} value
 * @return {void}
 */
SummaryRanges.prototype.addNum = function (value) {
  this.arr.add(value)
}

/**
 * @return {number[][]}
 */
SummaryRanges.prototype.getIntervals = function () {
  const intervals = []
  let left, right
  left = right = -1
  for (const value of Array.from(this.arr).sort((a, b) => a - b)) {
    if (left < 0) {
      left = right = value
    } else if (right + 1 === value) {
      right += 1
    } else {
      intervals.push([left, right])
      left = right = value
    }
  }
  intervals.push([left, right])
  return intervals
}

/**
 * Your SummaryRanges object will be instantiated and called as such:
 * var obj = new SummaryRanges()
 * obj.addNum(value)
 * var param_2 = obj.getIntervals()
 */

File: 투포인터/leetcode352(Hard)-Data Stream as Disjoint Intervals.py
# 383ms(37.46%), 19MB(75.92%)
class SummaryRanges:

    def __init__(self):
        self.arr = set()

    def addNum(self, value: int) -> None:
        self.arr.add(value)

    def getIntervals(self) -> List[List[int]]:
        intervals = []
        left = right = -1
        for value in sorted(list(self.arr)):
            if left < 0:
                left = right = value
            elif right + 1 == value:
                right = value # 혹은 right += 1
            else:
                intervals.append((left, right))
                left = right = value
        intervals.append((left, right))
        return intervals


# Your SummaryRanges object will be instantiated and called as such:
# obj = SummaryRanges()
# obj.addNum(value)
# param_2 = obj.getIntervals()